File: leetcode/0036-valid-sudoku/solution.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        set<char> colSet[9];
        set<char> rowSet[9];
        set<char> boxSet[9];
        
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                char ch = board[i][j];
                
                if (ch == '.')
                    continue;
                
                if (colSet[j].find(ch) != colSet[j].end())
                    return false;
                
                colSet[j].insert(ch);
                
                if (rowSet[i].find(ch) != rowSet[i].end())
                    return false;
                
                rowSet[i].insert(ch);
                
                int boxId = i / 3 * 3 + j / 3;
                
                if (boxSet[boxId].find(ch) != boxSet[boxId].end())
                    return false;
                
                boxSet[boxId].insert(ch);
            }
        }
        
        return true;
    }
};

File: leetcode/0036-valid-sudoku/solution.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        colSet = [set() for _ in range(9)]
        rowSet = [set() for _ in range(9)]
        boxSet = [set() for _ in range(9)]
        
        for i in range(9):
            for j in range(9):
                ch = board[i][j]
                
                if ch == '.':
                    continue
                
                if ch in colSet[j]:
                    return False
                
                colSet[j].add(ch)
                
                if ch in rowSet[i]:
                    return False
                
                rowSet[i].add(ch)
                
                boxId = i // 3 * 3 + j // 3
                
                if ch in boxSet[boxId]:
                    return False
                
                boxSet[boxId].add(ch)
        
        return True